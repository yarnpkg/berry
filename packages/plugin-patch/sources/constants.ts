// We use this for the patches to be regenerated without bumping the whole
// cache, like when the libzip had incorrect mtime in some cases
export const CACHE_VERSION = 3;
