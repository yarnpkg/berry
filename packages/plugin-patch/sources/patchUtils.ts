import {Cache, structUtils, Locator, Descriptor, Ident, Project, ThrowReport, miscUtils, FetchOptions, Package, execUtils} from '@yarnpkg/core';
import {npath, PortablePath, xfs, ppath, Filename, NativePath, CwdFS}                                                      from '@yarnpkg/fslib';

import {Hooks as PatchHooks}                                                                                               from './index';

export {applyPatchFile} from './tools/apply';
export {parsePatchFile} from './tools/parse';

const BUILTIN_REGEXP = /^builtin<([^>]+)>$/;

function parseSpec<T>(spec: string, sourceParser: (source: string) => T) {
  const {source, selector, params} = structUtils.parseRange(spec);
  if (source === null)
    throw new Error(`Patch locators must explicitly define their source`);

  const patchPaths = selector
    ? selector.split(/&/).map(path => npath.toPortablePath(path))
    : [];

  const parentLocator = params && typeof params.locator === `string`
    ? structUtils.parseLocator(params.locator)
    : null;

  const sourceVersion = params && typeof params.version === `string`
    ? params.version
    : null;

  const sourceItem = sourceParser(source);

  return {parentLocator, sourceItem, patchPaths, sourceVersion};
}

export function parseDescriptor(descriptor: Descriptor) {
  const {sourceItem, ...rest} = parseSpec(descriptor.range, structUtils.parseDescriptor);
  return {...rest, sourceDescriptor: sourceItem};
}

export function parseLocator(locator: Locator) {
  const {sourceItem, ...rest} = parseSpec(locator.reference, structUtils.parseLocator);
  return {...rest, sourceLocator: sourceItem};
}

function makeSpec<T>({parentLocator, sourceItem, patchPaths, sourceVersion, patchHash}: {parentLocator: Locator | null, sourceItem: T, patchPaths: Array<PortablePath>, sourceVersion?: string | null, patchHash?: string}, sourceStringifier: (source: T) => string) {
  const parentLocatorSpread = parentLocator !== null
    ? {locator: structUtils.stringifyLocator(parentLocator)}
    : {} as {};

  const sourceVersionSpread = typeof sourceVersion !== `undefined`
    ? {version: sourceVersion}
    : {} as {};

  const patchHashSpread = typeof patchHash !== `undefined`
    ? {hash: patchHash}
    : {} as {};

  return structUtils.makeRange({
    protocol: `patch:`,
    source: sourceStringifier(sourceItem),
    selector: patchPaths.join(`&`),
    params: {
      ...sourceVersionSpread,
      ...patchHashSpread,
      ...parentLocatorSpread,
    },
  });
}

export function makeDescriptor(ident: Ident, {parentLocator, sourceDescriptor, patchPaths}: ReturnType<typeof parseDescriptor>) {
  return structUtils.makeLocator(ident, makeSpec({parentLocator, sourceItem: sourceDescriptor, patchPaths}, structUtils.stringifyDescriptor));
}

export function makeLocator(ident: Ident, {parentLocator, sourcePackage, patchPaths, patchHash}: Omit<ReturnType<typeof parseLocator>, 'sourceLocator' | 'sourceVersion'> & {sourcePackage: Package, patchHash: string}) {
  return structUtils.makeLocator(ident, makeSpec({parentLocator, sourceItem: sourcePackage, sourceVersion: sourcePackage.version, patchPaths, patchHash}, structUtils.stringifyLocator));
}

type VisitPatchPathOptions<T> = {
  onAbsolute: (p: PortablePath) => T,
  onRelative: (p: PortablePath) => T,
  onBuiltin: (name: string) => T,
};

function visitPatchPath<T>({onAbsolute, onRelative, onBuiltin}: VisitPatchPathOptions<T>, patchPath: PortablePath) {
  const optional = patchPath.startsWith(`~`);
  if (optional)
    patchPath = patchPath.slice(1) as PortablePath;

  const builtinMatch = patchPath.match(BUILTIN_REGEXP);
  if (builtinMatch !== null)
    return onBuiltin(builtinMatch[1]);

  if (ppath.isAbsolute(patchPath)) {
    return onAbsolute(patchPath);
  } else {
    return onRelative(patchPath);
  }
}

export function extractPatchFlags(patchPath: PortablePath) {
  const optional = patchPath.startsWith(`~`);
  if (optional)
    patchPath = patchPath.slice(1) as PortablePath;

  return {optional};
}

export function isParentRequired(patchPath: PortablePath) {
  return visitPatchPath({
    onAbsolute: () => false,
    onRelative: () => true,
    onBuiltin: () => false,
  }, patchPath);
}

export async function loadPatchFiles(parentLocator: Locator | null, patchPaths: Array<PortablePath>, opts: FetchOptions) {
  // When the patch files use absolute paths we can directly access them via
  // their location on the disk. Otherwise we must go through the package fs.
  const parentFetch = parentLocator !== null
    ? await opts.fetcher.fetch(parentLocator, opts)
    : null;

  // If the package fs publicized its "original location" (for example like
  // in the case of "file:" packages), we use it to derive the real location.
  const effectiveParentFetch = parentFetch && parentFetch.localPath
    ? {packageFs: new CwdFS(PortablePath.root), prefixPath: ppath.relative(PortablePath.root, parentFetch.localPath)}
    : parentFetch;

  // Discard the parent fs unless we really need it to access the files
  if (parentFetch && parentFetch !== effectiveParentFetch && parentFetch.releaseFs)
    parentFetch.releaseFs();

  // First we obtain the specification for all the patches that we'll have to
  // apply to the original package.
  const patchFiles = await miscUtils.releaseAfterUseAsync(async () => {
    return await Promise.all(patchPaths.map(async patchPath => {
      const flags = extractPatchFlags(patchPath);

      const source = await visitPatchPath({
        onAbsolute: async () => {
          return await xfs.readFilePromise(patchPath, `utf8`);
        },

        onRelative: async () => {
          if (effectiveParentFetch === null)
            throw new Error(`Assertion failed: The parent locator should have been fetched`);

          return await effectiveParentFetch.packageFs.readFilePromise(ppath.join(effectiveParentFetch.prefixPath, patchPath), `utf8`);
        },

        onBuiltin: async name => {
          return await opts.project.configuration.firstHook((hooks: PatchHooks) => {
            return hooks.getBuiltinPatch;
          }, opts.project, name);
        },
      }, patchPath);

      return {...flags, source};
    }));
  });

  // Normalizes the line endings to prevent mismatches when cloning a
  // repository on Windows systems (the default settings for Git are to
  // convert newlines back and forth, which would mess with the checksum)
  for (const spec of patchFiles)
    if (typeof spec.source === `string`)
      spec.source = spec.source.replace(/\r\n?/g, `\n`);

  return patchFiles;
}

export async function extractPackageToDisk(locator: Locator, {cache, project}: {cache: Cache, project: Project}) {
  const pkg = project.storedPackages.get(locator.locatorHash);
  if (typeof pkg === `undefined`)
    throw new Error(`Assertion failed: Expected the package to be registered`);

  const checksums = project.storedChecksums;
  const report = new ThrowReport();

  const fetcher = project.configuration.makeFetcher();
  const fetchResult = await fetcher.fetch(locator, {cache, project, fetcher, checksums, report});

  const temp = await xfs.mktempPromise();

  const sourcePath = ppath.join(temp, `source` as Filename);
  const userPath = ppath.join(temp, `user` as Filename);
  const metaPath = ppath.join(temp, `.yarn-patch.json` as Filename);

  await Promise.all([
    xfs.copyPromise(sourcePath, fetchResult.prefixPath, {
      baseFs: fetchResult.packageFs,
    }),
    xfs.copyPromise(userPath, fetchResult.prefixPath, {
      baseFs: fetchResult.packageFs,
    }),
    xfs.writeJsonPromise(metaPath, {
      locator: structUtils.stringifyLocator(locator),
      version: pkg.version,
    }),
  ]);

  xfs.detachTemp(temp);
  return userPath;
}

export async function diffFolders(folderA: PortablePath, folderB: PortablePath) {
  const folderAN = npath.fromPortablePath(folderA).replace(/\\/g, `/`);
  const folderBN = npath.fromPortablePath(folderB).replace(/\\/g, `/`);

  const {stdout, stderr} = await execUtils.execvp(`git`, [`-c`, `core.safecrlf=false`, `diff`, `--src-prefix=a/`, `--dst-prefix=b/`, `--ignore-cr-at-eol`, `--full-index`, `--no-index`, `--text`, folderAN, folderBN], {
    cwd: npath.toPortablePath(process.cwd()),
    env: {
      ...process.env,
      //#region Predictable output
      // These variables aim to ignore the global git config so we get predictable output
      // https://git-scm.com/docs/git#Documentation/git.txt-codeGITCONFIGNOSYSTEMcode
      GIT_CONFIG_NOSYSTEM: `1`,
      HOME: ``,
      XDG_CONFIG_HOME: ``,
      USERPROFILE: ``,
      //#endregion
    },
  });

  // we cannot rely on exit code, because --no-index implies --exit-code
  // i.e. git diff will exit with 1 if there were differences
  if (stderr.length > 0)
    throw new Error(`Unable to diff directories. Make sure you have a recent version of 'git' available in PATH.\nThe following error was reported by 'git':\n${stderr}`);


  const normalizePath = folderAN.startsWith(`/`)
    ? (p: NativePath) => p.slice(1)
    : (p: NativePath) => p;

  return stdout
    .replace(new RegExp(`(a|b)(${miscUtils.escapeRegExp(`/${normalizePath(folderAN)}/`)})`, `g`), `$1/`)
    .replace(new RegExp(`(a|b)${miscUtils.escapeRegExp(`/${normalizePath(folderBN)}/`)}`, `g`), `$1/`)
    .replace(new RegExp(miscUtils.escapeRegExp(`${folderAN}/`), `g`), ``)
    .replace(new RegExp(miscUtils.escapeRegExp(`${folderBN}/`), `g`), ``);
}
