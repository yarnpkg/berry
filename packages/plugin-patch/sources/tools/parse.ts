import {miscUtils}           from '@yarnpkg/core';
import {PortablePath, npath} from '@yarnpkg/fslib';

const HEADER_REGEXP = /^@@ -(\d+)(,(\d+))? \+(\d+)(,(\d+))? @@.*/;

export type HunkHeader = {
  original: {
    start: number,
    length: number,
  },
  patched: {
    start: number,
    length: number,
  },
};

export function parseHunkHeaderLine(headerLine: string): HunkHeader {
  const match = headerLine.trim().match(HEADER_REGEXP);
  if (!match)
    throw new Error(`Bad header line: '${headerLine}'`);

  return {
    original: {
      start: Math.max(Number(match[1]), 1),
      length: Number(match[3] || 1),
    },
    patched: {
      start: Math.max(Number(match[4]), 1),
      length: Number(match[6] || 1),
    },
  };
}

export const NON_EXECUTABLE_FILE_MODE = 0o644;
export const EXECUTABLE_FILE_MODE = 0o755;

type FileMode =
  | typeof NON_EXECUTABLE_FILE_MODE
  | typeof EXECUTABLE_FILE_MODE;

type PatchMutationPart = {
  type: `context` | `insertion` | `deletion`,
  lines: Array<string>,
  noNewlineAtEndOfFile: boolean,
};

type FileRename = {
  type: `rename`,
  semverExclusivity: string | null,
  fromPath: PortablePath,
  toPath: PortablePath,
};

type FileModeChange = {
  type: `mode change`,
  semverExclusivity: string | null,
  path: PortablePath,
  oldMode: FileMode,
  newMode: FileMode,
};

export type FilePatch = {
  type: `patch`,
  semverExclusivity: string | null,
  path: PortablePath,
  hunks: Array<Hunk>,
  beforeHash: string | null,
  afterHash: string | null,
};

type FileDeletion = {
  type: `file deletion`
  semverExclusivity: string | null,
  path: PortablePath,
  mode: FileMode,
  hunk: Hunk | null,
  hash: string | null,
};

type FileCreation = {
  type: `file creation`,
  semverExclusivity: string | null,
  mode: FileMode,
  path: PortablePath,
  hunk: Hunk | null,
  hash: string | null,
};

export type PatchFilePart =
  | FilePatch
  | FileDeletion
  | FileCreation
  | FileRename
  | FileModeChange;

export type ParsedPatchFile =
  Array<PatchFilePart>;

type State =
  | `parsing header`
  | `parsing hunks`;

type FileDeets = {
  semverExclusivity: string | null,
  diffLineFromPath: string | null,
  diffLineToPath: string | null,
  oldMode: string | null,
  newMode: string | null,
  deletedFileMode: string | null,
  newFileMode: string | null,
  renameFrom: string | null,
  renameTo: string | null,
  beforeHash: string | null,
  afterHash: string | null,
  fromPath: string | null,
  toPath: string | null,
  hunks: Array<Hunk> | null,
};

export type Hunk = {
  header: HunkHeader,
  parts: PatchMutationPart[],
};

const emptyFilePatch = (): FileDeets => ({
  semverExclusivity: null,
  diffLineFromPath: null,
  diffLineToPath: null,
  oldMode: null,
  newMode: null,
  deletedFileMode: null,
  newFileMode: null,
  renameFrom: null,
  renameTo: null,
  beforeHash: null,
  afterHash: null,
  fromPath: null,
  toPath: null,
  hunks: null,
});

const emptyHunk = (headerLine: string): Hunk => ({
  header: parseHunkHeaderLine(headerLine),
  parts: [],
});

const hunkLinetypes: {[k: string]: PatchMutationPart['type'] | `pragma` | `header`} = {
  [`@`]: `header`,
  [`-`]: `deletion`,
  [`+`]: `insertion`,
  [` `]: `context`,
  [`\\`]: `pragma`,
  // Treat blank lines as context
  undefined: `context`,
};

function parsePatchLines(lines: Array<string>, {supportLegacyDiffs}: {supportLegacyDiffs: boolean}) {
  const result: Array<FileDeets> = [];

  let currentFilePatch: FileDeets = emptyFilePatch();
  let state: State = `parsing header`;
  let currentHunk: Hunk | null = null;
  let currentHunkMutationPart: PatchMutationPart | null = null;

  function commitHunk() {
    if (currentHunk) {
      if (currentHunkMutationPart) {
        currentHunk.parts.push(currentHunkMutationPart);
        currentHunkMutationPart = null;
      }

      currentFilePatch.hunks!.push(currentHunk);
      currentHunk = null;
    }
  }

  function commitFilePatch() {
    commitHunk();

    result.push(currentFilePatch);
    currentFilePatch = emptyFilePatch();
  }

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (state === `parsing header`) {
      if (line.startsWith(`@@`)) {
        state = `parsing hunks`;
        currentFilePatch.hunks = [];
        i -= 1;
      } else if (line.startsWith(`diff --git `)) {
        if (currentFilePatch && currentFilePatch.diffLineFromPath)
          commitFilePatch();

        const match = line.match(/^diff --git a\/(.*?) b\/(.*?)\s*$/);
        if (!match)
          throw new Error(`Bad diff line: ${line}`);

        currentFilePatch.diffLineFromPath = match[1];
        currentFilePatch.diffLineToPath = match[2];
      } else if (line.startsWith(`old mode `)) {
        currentFilePatch.oldMode = line.slice(`old mode `.length).trim();
      } else if (line.startsWith(`new mode `)) {
        currentFilePatch.newMode = line.slice(`new mode `.length).trim();
      } else if (line.startsWith(`deleted file mode `)) {
        currentFilePatch.deletedFileMode = line.slice(`deleted file mode `.length).trim();
      } else if (line.startsWith(`new file mode `)) {
        currentFilePatch.newFileMode = line.slice("new file mode ".length).trim();
      } else if (line.startsWith(`rename from `)) {
        currentFilePatch.renameFrom = line.slice(`rename from `.length).trim();
      } else if (line.startsWith(`rename to `)) {
        currentFilePatch.renameTo = line.slice(`rename to `.length).trim();
      } else if (line.startsWith(`index `)) {
        const match = line.match(/(\w+)\.\.(\w+)/);
        if (!match)
          continue;

        currentFilePatch.beforeHash = match[1];
        currentFilePatch.afterHash = match[2];
      } else if (line.startsWith(`semver exclusivity `)) {
        currentFilePatch.semverExclusivity = line.slice(`semver exclusivity `.length).trim();
      } else if (line.startsWith(`--- `)) {
        currentFilePatch.fromPath = line.slice(`--- a/`.length).trim();
      } else if (line.startsWith(`+++ `)) {
        currentFilePatch.toPath = line.slice(`+++ b/`.length).trim();
      }
    } else {
      if (supportLegacyDiffs && line.startsWith(`--- a/`)) {
        state = `parsing header`;
        commitFilePatch();
        i -= 1;
        continue;
      }

      // parsing hunks
      const lineType = hunkLinetypes[line[0]] || null;
      switch (lineType) {
        case `header`: {
          commitHunk();
          currentHunk = emptyHunk(line);
        } break;

        case null: {
          // unrecognized, bail out
          state = `parsing header`;
          commitFilePatch();
          i -= 1;
        } break;

        case `pragma`: {
          if (!line.startsWith("\\ No newline at end of file"))
            throw new Error(`Unrecognized pragma in patch file: ${line}`);

          if (!currentHunkMutationPart)
            throw new Error(`Bad parser state: No newline at EOF pragma encountered without context`);

          currentHunkMutationPart.noNewlineAtEndOfFile = true;
        } break;

        case `insertion`:
        case `deletion`:
        case `context`: {
          if (!currentHunk)
            throw new Error(`Bad parser state: Hunk lines encountered before hunk header`);

          if (currentHunkMutationPart && currentHunkMutationPart.type !== lineType) {
            currentHunk.parts.push(currentHunkMutationPart);
            currentHunkMutationPart = null;
          }

          if (!currentHunkMutationPart) {
            currentHunkMutationPart = {
              type: lineType,
              lines: [],
              noNewlineAtEndOfFile: false,
            };
          }

          currentHunkMutationPart.lines.push(line.slice(1));
        } break;

        default: {
          miscUtils.assertNever(lineType);
        } break;
      }
    }
  }

  commitFilePatch();

  for (const {hunks} of result)
    if (hunks)
      for (const hunk of hunks)
        verifyHunkIntegrity(hunk);

  return result;
}

export function interpretParsedPatchFile(files: Array<FileDeets>): ParsedPatchFile {
  const result: ParsedPatchFile = [];

  for (const file of files) {
    const {
      semverExclusivity,
      diffLineFromPath,
      diffLineToPath,
      oldMode,
      newMode,
      deletedFileMode,
      newFileMode,
      renameFrom,
      renameTo,
      beforeHash,
      afterHash,
      fromPath,
      toPath,
      hunks,
    } = file;

    const type: PatchFilePart["type"] = renameFrom
      ? `rename`
      : deletedFileMode
        ? `file deletion`
        : newFileMode
          ? `file creation`
          : hunks && hunks.length > 0
            ? `patch`
            : `mode change`;

    let destinationFilePath: string | null = null;
    switch (type) {
      case `rename`: {
        if (!renameFrom || !renameTo)
          throw new Error(`Bad parser state: rename from & to not given`);

        result.push({
          type: `rename`,
          semverExclusivity,
          fromPath: npath.toPortablePath(renameFrom),
          toPath: npath.toPortablePath(renameTo),
        });

        destinationFilePath = renameTo;
      } break;

      case `file deletion`: {
        const path = diffLineFromPath || fromPath;
        if (!path)
          throw new Error("Bad parse state: no path given for file deletion");

        result.push({
          type: `file deletion`,
          semverExclusivity,
          hunk: (hunks && hunks[0]) || null,
          path: npath.toPortablePath(path),
          mode: parseFileMode(deletedFileMode!),
          hash: beforeHash,
        });
      } break;

      case `file creation`: {
        const path = diffLineToPath || toPath;
        if (!path)
          throw new Error(`Bad parse state: no path given for file creation`);

        result.push({
          type: `file creation`,
          semverExclusivity,
          hunk: (hunks && hunks[0]) || null,
          path: npath.toPortablePath(path),
          mode: parseFileMode(newFileMode!),
          hash: afterHash,
        });
      } break;

      case `patch`:
      case `mode change`: {
        destinationFilePath = toPath || diffLineToPath;
      } break;

      default: {
        miscUtils.assertNever(type);
      } break;
    }

    if (destinationFilePath && oldMode && newMode && oldMode !== newMode) {
      result.push({
        type: `mode change`,
        semverExclusivity,
        path: npath.toPortablePath(destinationFilePath),
        oldMode: parseFileMode(oldMode),
        newMode: parseFileMode(newMode),
      });
    }

    if (destinationFilePath && hunks && hunks.length) {
      result.push({
        type: `patch`,
        semverExclusivity,
        path: npath.toPortablePath(destinationFilePath),
        hunks,
        beforeHash,
        afterHash,
      });
    }
  }

  return result;
}

function parseFileMode(mode: string): FileMode {
  const parsedMode = parseInt(mode, 8) & 0o777;

  if (parsedMode !== NON_EXECUTABLE_FILE_MODE && parsedMode !== EXECUTABLE_FILE_MODE)
    throw new Error(`Unexpected file mode string: ${mode}`);

  return parsedMode;
}

export function parsePatchFile(file: string): ParsedPatchFile {
  const lines = file.split(/\n/g);
  if (lines[lines.length - 1] === ``)
    lines.pop();

  try {
    return interpretParsedPatchFile(parsePatchLines(lines, {supportLegacyDiffs: false}));
  } catch (e) {
    if (e instanceof Error && e.message === `hunk header integrity check failed`) {
      return interpretParsedPatchFile(parsePatchLines(lines, {supportLegacyDiffs: true}));
    } else {
      throw e;
    }
  }
}

export function verifyHunkIntegrity(hunk: Hunk) {
  // verify hunk integrity
  let originalLength = 0;
  let patchedLength = 0;

  for (const {type, lines} of hunk.parts) {
    switch (type) {
      case `context`: {
        patchedLength += lines.length;
        originalLength += lines.length;
      } break;

      case "deletion": {
        originalLength += lines.length;
      } break;

      case "insertion": {
        patchedLength += lines.length;
      } break;

      default: {
        miscUtils.assertNever(type);
      } break;
    }
  }

  if (originalLength !== hunk.header.original.length || patchedLength !== hunk.header.patched.length) {
    throw new Error(`hunk header integrity check failed`);
  }
}
