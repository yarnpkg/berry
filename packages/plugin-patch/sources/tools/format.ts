import {miscUtils, Report, formatUtils, MessageName, Configuration} from '@yarnpkg/core';

import {Hunk, PatchMutationType}                                    from './parse';

export function reportHunk(hunk: Hunk, {configuration, report}: {configuration: Configuration, report: Report}) {
  for (const part of hunk.parts) {
    for (const line of part.lines) {
      switch (part.type) {
        case PatchMutationType.Context: {
          report.reportInfo(null, `  ${formatUtils.pretty(configuration, line, `grey`)}`);
        } break;

        case PatchMutationType.Deletion: {
          report.reportError(MessageName.FROZEN_LOCKFILE_EXCEPTION, `- ${formatUtils.pretty(configuration, line, formatUtils.Type.REMOVED)}`);
        } break;

        case PatchMutationType.Insertion: {
          report.reportError(MessageName.FROZEN_LOCKFILE_EXCEPTION, `+ ${formatUtils.pretty(configuration, line, formatUtils.Type.ADDED)}`);
        } break;

        default: {
          miscUtils.assertNever(part.type);
        }
      }
    }
  }
}
