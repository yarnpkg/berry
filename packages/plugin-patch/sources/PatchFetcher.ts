import {Fetcher, FetchOptions, MinimalFetchOptions, ReportError, MessageName, Report} from '@yarnpkg/core';
import {Locator}                                                                      from '@yarnpkg/core';
import {miscUtils, structUtils}                                                       from '@yarnpkg/core';
import {ppath, xfs, ZipFS, Filename, CwdFS, PortablePath}                             from '@yarnpkg/fslib';
import {getLibzipPromise}                                                             from '@yarnpkg/libzip';

import * as patchUtils                                                                from './patchUtils';
import {UnmatchedHunkError}                                                           from './tools/UnmatchedHunkError';
import {reportHunk}                                                                   from './tools/format';

export class PatchFetcher implements Fetcher {
  supports(locator: Locator, opts: MinimalFetchOptions) {
    if (!locator.reference.startsWith(`patch:`))
      return false;

    return true;
  }

  getLocalPath(locator: Locator, opts: FetchOptions) {
    return null;
  }

  async fetch(locator: Locator, opts: FetchOptions) {
    const expectedChecksum = opts.checksums.get(locator.locatorHash) || null;

    const [packageFs, releaseFs, checksum] = await opts.cache.fetchPackageFromCache(locator, expectedChecksum, {
      onHit: () => opts.report.reportCacheHit(locator),
      onMiss: () => opts.report.reportCacheMiss(locator, `${structUtils.prettyLocator(opts.project.configuration, locator)} can't be found in the cache and will be fetched from the disk`),
      loader: () => this.patchPackage(locator, opts),
      skipIntegrityCheck: opts.skipIntegrityCheck,
    });

    return {
      packageFs,
      releaseFs,
      prefixPath: structUtils.getIdentVendorPath(locator),
      localPath: this.getLocalPath(locator, opts),
      checksum,
    };
  }

  private async patchPackage(locator: Locator, opts: FetchOptions) {
    const {parentLocator, sourceLocator, sourceVersion, patchPaths} = patchUtils.parseLocator(locator);
    const patchFiles = await patchUtils.loadPatchFiles(parentLocator, patchPaths, opts);

    const tmpDir = await xfs.mktempPromise();
    const currentFile = ppath.join(tmpDir, `current.zip` as Filename);

    const sourceFetch = await opts.fetcher.fetch(sourceLocator, opts);
    const prefixPath = structUtils.getIdentVendorPath(locator);

    const libzip = await getLibzipPromise();

    // First we create a copy of the package that we'll be free to mutate
    const initialCopy = new ZipFS(currentFile, {
      libzip,
      create: true,
      level: opts.project.configuration.get(`compressionLevel`),
    });

    await initialCopy.mkdirpPromise(prefixPath);

    await miscUtils.releaseAfterUseAsync(async () => {
      await initialCopy.copyPromise(prefixPath, sourceFetch.prefixPath, {baseFs: sourceFetch.packageFs, stableSort: true});
    }, sourceFetch.releaseFs);

    initialCopy.saveAndClose();

    for (const {source, optional} of patchFiles) {
      if (source === null)
        continue;

      // Then for each patchfile, we open this copy anew, and try to apply the
      // changeset. We need to open it for each patchfile (rather than only a
      // single time) because it lets us easily rollback when hitting errors
      // on optional patches (we just need to call `discardAndClose`).
      const patchedPackage = new ZipFS(currentFile, {
        libzip,
        level: opts.project.configuration.get(`compressionLevel`),
      });

      const patchFs = new CwdFS(ppath.resolve(PortablePath.root, prefixPath), {
        baseFs: patchedPackage,
      });

      try {
        await patchUtils.applyPatchFile(patchUtils.parsePatchFile(source), {
          baseFs: patchFs,
          version: sourceVersion,
        });
      } catch (err) {
        if (!(err instanceof UnmatchedHunkError))
          throw err;

        const enableInlineHunks = opts.project.configuration.get(`enableInlineHunks`);
        const suggestion = !enableInlineHunks && !optional
          ? ` (set enableInlineHunks for details)`
          : ``;

        const message = `${structUtils.prettyLocator(opts.project.configuration, locator)}: ${err.message}${suggestion}`;
        const reportExtra = (report: Report) => {
          if (!enableInlineHunks)
            return;

          reportHunk(err.hunk, {
            configuration: opts.project.configuration,
            report,
          });
        };

        // By discarding the current changes, the next patch will start from
        // where we were.
        patchedPackage.discardAndClose();

        if (optional) {
          opts.report.reportWarningOnce(MessageName.PATCH_HUNK_FAILED, message, {reportExtra});
          continue;
        } else {
          throw new ReportError(MessageName.PATCH_HUNK_FAILED, message, reportExtra);
        }
      }

      patchedPackage.saveAndClose();
    }

    return new ZipFS(currentFile, {
      libzip,
      level: opts.project.configuration.get(`compressionLevel`),
    });
  }
}
