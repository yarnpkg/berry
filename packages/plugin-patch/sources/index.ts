import {Plugin, Project, SettingsType} from '@yarnpkg/core';
import {PortablePath}                  from '@yarnpkg/fslib';

import {PatchFetcher}                  from './PatchFetcher';
import {PatchResolver}                 from './PatchResolver';
import PatchCommit                     from './commands/patchCommit';
import Patch                           from './commands/patch';
import * as patchUtils                 from './patchUtils';

export {patchUtils};

export interface Hooks {
  getBuiltinPatch?: (
    project: Project,
    name: string,
  ) => Promise<string | null | void>,
}

declare module '@yarnpkg/core' {
  interface ConfigurationValueMap {
    enableInlineHunks: boolean;
    patchFolder: PortablePath;
  }
}

const plugin: Plugin = {
  configuration: {
    enableInlineHunks: {
      description: `If true, the installs will print unmatched patch hunks`,
      type: SettingsType.BOOLEAN,
      default: false,
    },
    patchFolder: {
      description: `Folder where the patch files must be written`,
      type: SettingsType.ABSOLUTE_PATH,
      default: `./.yarn/patches`,
    },
  },
  commands: [
    PatchCommit,
    Patch,
  ],
  fetchers: [
    PatchFetcher,
  ],
  resolvers: [
    PatchResolver,
  ],
};

// eslint-disable-next-line arca/no-default-export
export default plugin;
