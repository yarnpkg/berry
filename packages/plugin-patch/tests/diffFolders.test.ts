import {npath, NodeFS}  from '@yarnpkg/fslib';

import {diffFolders}    from '../sources/patchUtils';
import {parsePatchFile} from '../sources/tools/parse';

describe(`diffFolders`,  () => {
  const fs = new NodeFS();
  const fixtures = npath.join(__dirname, `fixtures`);

  for (const iterator of fs.readdirSync(npath.toPortablePath(fixtures))) {
    it(`Makes and parses diff for '${iterator}'`, async () => {
      const diff = await diffFolders(
        npath.toPortablePath(npath.join(fixtures, iterator, `a`)),
        npath.toPortablePath(npath.join(fixtures, iterator, `b`))
      );

      expect(diff).toMatchSnapshot();
      expect(parsePatchFile(diff)).toMatchSnapshot();
    });
  }
});
