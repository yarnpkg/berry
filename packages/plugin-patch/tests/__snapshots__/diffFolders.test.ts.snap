// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffFolders Makes and parses diff for 'create' 1`] = `
"diff --git a/foo.txt b/foo.txt
new file mode 100644
index 0000000000000000000000000000000000000000..257cc5642cb1a054f08cc83f2d943e56fd3ebe99
--- /dev/null
+++ b/foo.txt
@@ -0,0 +1 @@
+foo
"
`;

exports[`diffFolders Makes and parses diff for 'create' 2`] = `
Array [
  Object {
    "hash": "257cc5642cb1a054f08cc83f2d943e56fd3ebe99",
    "hunk": Object {
      "header": Object {
        "original": Object {
          "length": 0,
          "start": 1,
        },
        "patched": Object {
          "length": 1,
          "start": 1,
        },
      },
      "parts": Array [
        Object {
          "lines": Array [
            "foo",
          ],
          "noNewlineAtEndOfFile": false,
          "type": "insertion",
        },
      ],
    },
    "mode": 420,
    "path": "foo.txt",
    "semverExclusivity": null,
    "type": "file creation",
  },
]
`;

exports[`diffFolders Makes and parses diff for 'remove' 1`] = `
"diff --git a/foo.txt b/foo.txt
deleted file mode 100644
index 257cc5642cb1a054f08cc83f2d943e56fd3ebe99..0000000000000000000000000000000000000000
--- a/foo.txt
+++ /dev/null
@@ -1 +0,0 @@
-foo
"
`;

exports[`diffFolders Makes and parses diff for 'remove' 2`] = `
Array [
  Object {
    "hash": "257cc5642cb1a054f08cc83f2d943e56fd3ebe99",
    "hunk": Object {
      "header": Object {
        "original": Object {
          "length": 1,
          "start": 1,
        },
        "patched": Object {
          "length": 0,
          "start": 1,
        },
      },
      "parts": Array [
        Object {
          "lines": Array [
            "foo",
          ],
          "noNewlineAtEndOfFile": false,
          "type": "deletion",
        },
      ],
    },
    "mode": 420,
    "path": "foo.txt",
    "semverExclusivity": null,
    "type": "file deletion",
  },
]
`;

exports[`diffFolders Makes and parses diff for 'rename' 1`] = `
"diff --git a/foo.txt b/foo-new.txt
similarity index 100%
rename from foo.txt
rename to foo-new.txt
"
`;

exports[`diffFolders Makes and parses diff for 'rename' 2`] = `
Array [
  Object {
    "fromPath": "foo.txt",
    "semverExclusivity": null,
    "toPath": "foo-new.txt",
    "type": "rename",
  },
]
`;

exports[`diffFolders Makes and parses diff for 'update' 1`] = `
"diff --git a/foo.txt b/foo.txt
index 257cc5642cb1a054f08cc83f2d943e56fd3ebe99..5716ca5987cbf97d6bb54920bea6adde242d87e6 100644
--- a/foo.txt
+++ b/foo.txt
@@ -1 +1 @@
-foo
+bar
"
`;

exports[`diffFolders Makes and parses diff for 'update' 2`] = `
Array [
  Object {
    "afterHash": "5716ca5987cbf97d6bb54920bea6adde242d87e6",
    "beforeHash": "257cc5642cb1a054f08cc83f2d943e56fd3ebe99",
    "hunks": Array [
      Object {
        "header": Object {
          "original": Object {
            "length": 1,
            "start": 1,
          },
          "patched": Object {
            "length": 1,
            "start": 1,
          },
        },
        "parts": Array [
          Object {
            "lines": Array [
              "foo",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "deletion",
          },
          Object {
            "lines": Array [
              "bar",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "insertion",
          },
        ],
      },
    ],
    "path": "foo.txt",
    "semverExclusivity": null,
    "type": "patch",
  },
]
`;
